### ISSUE TITLE
Add necessary but missing connectionLost method to client.Resolver

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5471]
|<img alt="BobNovas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| BobNovas reported|
|-|-|
|Trac ID|trac#5471|
|Type|defect|
|Created|2012-02-01 15:41:30Z|
|Branch|https://github.com/twisted/twisted/tree/dns-connectionlost-5471|

This patch fixes a problem that happens when a Resolver.queryUDP result is truncated and Resolver.filterAnswers calls Resolver.queryTCP. queryTCP passes self.factory, which is a DNSClientFactory, to reactor.connectTCP.  Eventually,  client.DNSClientFactory.buildProtocol passes self.controller to the TCP protocol. In DNSClientFactory, self.controller is the resolver, which didn't have a connectionLost method.  

This missing method (which this patch supplies) is called when the TCP protocol disconnects.

A typical traceback showing the problem is:
```
	  File "C:\Python26-x64\lib\site-packages\twisted-11.1.0-py2.6-win-amd64.egg\twisted\python\log.py", line 84, in callWithLogger
	    return callWithContext({"system": lp}, func, *args, **kw)
	  File "C:\Python26-x64\lib\site-packages\twisted-11.1.0-py2.6-win-amd64.egg\twisted\python\log.py", line 69, in callWithContext
	    return context.call({ILogContext: newCtx}, func, *args, **kw)
	  File "C:\Python26-x64\lib\site-packages\twisted-11.1.0-py2.6-win-amd64.egg\twisted\python\context.py", line 118, in callWithContext
	    return self.currentContext().callWithContext(ctx, func, *args, **kw)
	  File "C:\Python26-x64\lib\site-packages\twisted-11.1.0-py2.6-win-amd64.egg\twisted\python\context.py", line 81, in callWithContext
	    return func(*args,**kw)
	  File "C:\Python26-x64\lib\site-packages\twisted-11.1.0-py2.6-win-amd64.egg\twisted\internet\selectreactor.py", line 155, in _doReadOrWrite
	    self._disconnectSelectable(selectable, why, method=="doRead")
	  File "C:\Python26-x64\lib\site-packages\twisted-11.1.0-py2.6-win-amd64.egg\twisted\internet\posixbase.py", line 260, in _disconnectSelectable
	    selectable.readConnectionLost(f)
	  File "C:\Python26-x64\lib\site-packages\twisted-11.1.0-py2.6-win-amd64.egg\twisted\internet\tcp.py", line 257, in readConnectionLost
	    self.connectionLost(reason)
	  File "C:\Python26-x64\lib\site-packages\twisted-11.1.0-py2.6-win-amd64.egg\twisted\internet\tcp.py", line 433, in connectionLost
	    Connection.connectionLost(self, reason)
	  File "C:\Python26-x64\lib\site-packages\twisted-11.1.0-py2.6-win-amd64.egg\twisted\internet\tcp.py", line 277, in connectionLost
	    protocol.connectionLost(reason)
	  File "C:\Python26-x64\lib\site-packages\twisted-11.1.0-py2.6-win-amd64.egg\twisted\names\dns.py", line 1908, in connectionLost
	    self.controller.connectionLost(self)
	exceptions.AttributeError: Resolver instance has no attribute 'connectionLost'
```

You can cause this traceback by querying (without this patch) for any resource record that will cause a truncation error. These are rare, though they will become more common with the use of DNSSEC.

(It is incorrect to assume that the connectionLost method in DNSClientFactory will cover this problem. I don't think that method is ever called.)

Attachments:

* [04-fixResolverConnectionLostProblem.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/111/111c78a9861e51488b88b8014476fbb56f3c0566/214b0dbc2dd6a11b8a2e7f2df6a34c55c8cec69d.patch) (5106 bytes) - added by BobNovas on 2012-02-01 15:41:48Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5471 5471
type__defect defect
reporter__BobNovas BobNovas
priority__normal normal
milestone__ 
branch__branches_dns_connectionlost_5471 branches/dns-connectionlost-5471
branch_author__BobNovas__itamar BobNovas, itamar
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1328110890000000 1328110890000000
changetime__1347194817000000 1347194817000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="BobNovas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|BobNovas commented|
|-|-|

Note that when I say "You can cause this traceback by querying (without this patch) for any resource record that will cause a truncation error.", what I forgot to mention is that the exception will be raised and the traceback will occur several MINUTES after the TCP protocol successfully returns the answer, when the protocol disconnects.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @itamarst|
|-|-|

I'll try to take a look at this soon.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [33513]) Branching to 'dns-connectionlost-5471'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

I've updated the branch with a better test. The original test you wrote didn't actually test the problem. First, it didn't actually do what your original (very accurate) description said; it used far too many Fake objects. Second, it didn't actually test the fix -- if I commented out the new connectionLost method you added, the test still passed.

I've also added a news file, and updated the code to follow our coding standards.

http://twistedmatrix.com/trac/changeset?new=33521%40branches%2Fdns-connectionlost-5471&old=33511%40trunk

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst removed owner|
|-|-|

Ready for review.

(Almost finished) Buildbot run: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/dns-connectionlost-5471

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

The news file should be in `names/topfiles`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.

  1. We typically don't link to a ticket number in test method docstrings.  Basically every test method _could_ link to a ticket (except the really, really old tests).  Version control history is typically sufficient to produce that information and does so automatically rather than requiring manual steps.
  1. The test should call `makeConnection` instead of `connectionMade`, shouldn't it?
  1. The news fragment doesn't follow the style recommendations on ReviewProcess#Newsfiles

Apart from those issues and the location of the news fragment, looks good.  Please merge when those points are addressed.

**Comment by twisted-trac** [comment]
|<img alt="BobNovas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|BobNovas commented|
|-|-|

Beautiful. Thanks very much for fixing the test code - I had a hard time with that but I understand your code and I like your new test method. I didn't have a deep enough understanding of the code to write the correct test, though.  The underlying problem has been a real issue for me, as I use DNS to read big TXT records containing keys and they invariably wind up in the DNS TCP protocol due to size, and then throw an exception, which this should fix.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set status to `closed`|
|-|-|

(In [33527]) Merge dns-connectionlost-5471: DNS TCP protocols no longer raise exception when disconnecting.

Author: BobNovas, itamar
Review: itamar, exarkun
Fixes: [#5471](https://github.com/twisted/twisted/issues/5471)

When a TCP DNS protocol disconnects, it is now correctly unregistered from the Resolver it originated from.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_11"></a>|@wallrj commented|
|-|-|

This was a duplicate of [#4015](https://github.com/twisted/twisted/issues/4015)
